src/os/bluestore/BlueStore.cc

```diff
@@ -11056,6 +11073,14 @@ int BlueStore::_upgrade_super()
       int r = db->submit_transaction_sync(t);
       ceph_assert(r == 0);
     }
+    if (ondisk_format == 3) {
+      // changes:
+      // - FreelistManager keeps meta within bdev label
+      int r = _write_out_fm_meta(0);
+      ceph_assert(r == 0);
+
+      ondisk_format = 4;
+    }
```

Comment:
How about going to ondisk_format=4 only for duration of expansion procedure?
This would reduce a lot of corner cases for failed expanding / downgrading intermix.

Reply:
Not sure this is a good idea. First of all I'd like eventually to have FM parameters in device's label only. So from now on there primary location is in the label. While undoing format assumes this is a temporary location.
And to be honest the trick with setting label back and forth doesn't look clear and straightforward enough . Not to mention it differs from everything we had with ondisk_format before.